### ISSUE TITLE
minidump loading

### ISSUE THREAD

**Issue by s0i37** [angr/angr#1641]
Hello.
May you please help me. I try to load whole virtual memory of some application with minidump.
As I know, angr doesn't support minidump format. Therefore I make dump every memory page to disk through thirt-party tools and then I try to load these page dumps with angr like:
```
project = angr.Project( "/bin/true", load_options={'auto_load_libs':False, 'main_opts': {'backend': 'blob', 'custom_arch': 'x86', 'custom_base_addr': 0xffffffff} } )
state = project.factory.entry_state()
state.memory.store(0xADDR, memory, len(memory))
```
(btw how to create an empty project, without specify binary file?)
But during loading ~500 memory pages by various virtual addresses angr consumes whole my memory.

**Comment by ltfish** [comment]
This can make a good test case for our memory model - I'm surprised that writing 500 pages into SimSymbolicMemory can exhaust your memory.

Your best bet at this point (before we figure out what's going on and introduce any sort of optimization to handle this use case better) is to write a loader backend for CLE that loads a minidump and maps its data to a bunch of memory backers. It should be pretty straightforward to write such a backend. Python has [a library](https://github.com/skelsec/minidump) to parse minidumps.

**Comment by s0i37** [comment]
Am I understand right that function `state.memory.store()` write concretic memory?

**Comment by s0i37** [comment]
> (btw how to create an empty project, without specify binary file?)

project = angr.load_shellcode("\x90".encode(), 'i686', start_offset=0, load_address=0)

**Comment by ltfish** [comment]
> Am I understand right that function `state.memory.store()` write concretic memory?

I'm not sure what you mean by "concretic memory". It writes to the symbolic memory model that a `SimState` has. It can write both symbolic data and concrete data. Reading/writing symbolic data and concrete data are similarly expensive into angr's symbolic memory model because those accesses will follow along almost the same code path.

**Comment by nathanjackson** [comment]
@s0i37 @ltfish Did you either of you ever figure this one out? I am running into a similar issue, but not with minidumps. I admit my expectations might be too high.

I'm trying to load ~500mb of data using `state.memory.store(...)`. I quickly exhaust all 32GB of RAM on my workstation.

```
with open("angr_dump.bin", "rb") as dump_file:
    while True:
        try:
            region_base = struct.unpack("I", dump_file.read(4))[0]
            region_size = struct.unpack("Q", dump_file.read(8))[0]
        except:
            break
        print("region base = 0x%X" % (region_base))
        print("region size = %u" % (region_size))
        region_data = dump_file.read(region_size)

        state.memory.store(region_base, region_data)
```

I'd even be open to subclassing\implementing a new SimMemory plugin alternative that accesses the data from a memory mapped file if that's possible.

**Comment by nathanjackson** [comment]
It seems like the correct solution is to write a CLE loader instead of using `memory.store`... disregard my above message.

**Comment by rhelmot** [comment]
#209 is a contributed minudump CLE loader that I have been meaning to
review for a while. I will try to merge it today.

**Comment by rhelmot** [comment]
lol, I forgot what repository I was replying to. I merged angr/cle#209 today, so minidumps can now be directly loaded.
